### PR TITLE
NXP-20571: make platform type service retrieve subtype definitions from core

### PR DESCRIPTION
**Pull request by Gabez0r**

### CODE REVIEW

File: nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java [hunk 1]
```diff
@@ -242,7 +243,41 @@ public boolean hasType(String id) {
...
-        return types.get(id);
+        Type type = types.get(id);
+        if (type == null) {
+            return null;
+        } else {
+            Map<String, SubType> allowedSubTypes = getCoreAllowedSubtypes(type);
+            Type newType = type.clone();
+            newType.setAllowedSubTypes(allowedSubTypes);
+            newType.setDeniedSubTypes(new String[0]);
+            return newType;
```
Comment (atchertchian) [inline nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java]: maybe add a comment in the code here, explaining that denied sub types are already filtered from the allowed sub types by the core API (? if that's indeed the explanation :))

File: nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java [hunk 2]
```diff
@@ -242,7 +243,42 @@ public boolean hasType(String id) {
     }
 
     public Type getType(String id) {
-        return types.get(id);
+        Type type = types.get(id);
+        if (type == null) {
+            return null;
+        } else {
```
Comment (efge) [inline nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java]: This is early exit, so you can move the `else` part unindented below.

File: nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java [hunk 3]
```diff
@@ -242,7 +243,42 @@ public boolean hasType(String id) {
...
+     * @since 8.10
+     */
+    protected Map<String, SubType> getCoreAllowedSubtypes(Type type) {
+        SchemaManagerImpl schemaManager = (SchemaManagerImpl) Framework.getService(SchemaManager.class);
+        Collection<String> coreAllowedSubtypes = schemaManager.getAllowedSubTypes(type.getId());
+        Map<String, SubType> ecmSubTypes = type.getAllowedSubTypes();
+        Map<String, SubType> allowedSubTypes = new HashMap<>();
+
+        SubType subtype;
+        if (coreAllowedSubtypes != null) {
```
Comment (efge) [inline nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java]: Use early exit:

    if (coreAllowedSubtypes == null) {
        // there are no subtypes to take care of
        return Collections.emptyMap();
    }

Then the rest unindented. Also move the `allowedSubTypes` and `subtype` declarations closer to their actual use.

File: nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java [hunk 4]
```diff
@@ -242,7 +243,42 @@ public boolean hasType(String id) {
...
+            newType.setDeniedSubTypes(new String[0]);
+            return newType;
+        }
+    }
+
+    /**
+     * @since 8.10
+     */
+    protected Map<String, SubType> getCoreAllowedSubtypes(Type type) {
+        SchemaManagerImpl schemaManager = (SchemaManagerImpl) Framework.getService(SchemaManager.class);
```
Comment (efge) [inline nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java]: You should only need `SchemaManager` here. If necessary add `getAllowedSubTypes` to that interface (I can't check if it's already there at the moment).

File: nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java [hunk 5]
```diff
@@ -29,6 +29,7 @@
 import java.util.Map;
 import java.util.Set;
 
+import org.nuxeo.ecm.core.schema.DocumentType;
```
Comment (efge) [inline nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java]: I don't think this import is needed.

Comment (Gabez0r) [inline nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java]: It isn't. Will be removed.

File: nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java (hunk 2 above)
Comment (Gabez0r) [inline nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java]: Done.

File: nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java (hunk 4 above)
Comment (Gabez0r) [inline nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java]: Done.

File: nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java (hunk 3 above)
Comment (Gabez0r) [inline nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java]: Done.

**Comment by efge** [comment]
👍

File: nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeService.java [hunk 6]
```diff
@@ -57,13 +58,20 @@
 
     private TypeRegistry typeRegistry;
 
+    private Runnable recomputeCallback;
+
     @Override
     public void activate(ComponentContext context) {
         typeRegistry = new TypeRegistry();
+        recomputeCallback = () -> typeRegistry.recomputeTypes();
```
Comment (efge) [inline nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeService.java]: Or simply `recomputeCallback = typeRegistry::recomputeTypes`

File: nuxeo-core/nuxeo-core-schema/src/main/java/org/nuxeo/ecm/core/schema/SchemaManagerImpl.java [hunk 7]
```diff
@@ -779,4 +783,30 @@ public void flushPendingsRegistration() {
...
+     * @since 8.10
+     */
+    public void unregisterRecomputeCallback(Runnable callback) {
+        recomputeCallbacks.remove(callback);
+    }
+
+    /**
+     * @since 8.10
+     */
+    protected void executeCallbacks() {
```
Comment (efge) [inline nuxeo-core/nuxeo-core-schema/src/main/java/org/nuxeo/ecm/core/schema/SchemaManagerImpl.java]: Rename to `executeRecomputeCallbacks`

File: nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeService.java [hunk 8]
```diff
@@ -57,13 +58,20 @@
...
+        SchemaManagerImpl schemaManager = (SchemaManagerImpl)Framework.getService(SchemaManager.class);
```
Comment (efge) [inline nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeService.java]: Space after cast.

Comment (Gabez0r) [inline nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeService.java]: Done.

File: nuxeo-core/nuxeo-core-schema/src/main/java/org/nuxeo/ecm/core/schema/SchemaManagerImpl.java (hunk 7 above)
Comment (Gabez0r) [inline nuxeo-core/nuxeo-core-schema/src/main/java/org/nuxeo/ecm/core/schema/SchemaManagerImpl.java]: Done.

**Comment by efge** [comment]
👍

**Pull request closed without merging**